ord_schema/validations.py

```diff
@@ -496,6 +496,35 @@ def validate_reaction_input(message: reaction_pb2.ReactionInput):
                     ValidationWarning,
                 )
 
+    texture_type_to_state_of_matter = {
+        reaction_pb2.ProductCompound.Texture.UNSPECIFIED: None,
+        reaction_pb2.ProductCompound.Texture.CUSTOM: None,
+        reaction_pb2.ProductCompound.Texture.GAS: 1,
+        reaction_pb2.ProductCompound.Texture.OIL: 2,
+        reaction_pb2.ProductCompound.Texture.FOAM: 2,
+        reaction_pb2.ProductCompound.Texture.LIQUID: 2,
+        reaction_pb2.ProductCompound.Texture.POWDER: 3,
+        reaction_pb2.ProductCompound.Texture.CRYSTAL: 3,
+        reaction_pb2.ProductCompound.Texture.WAX: 3,
+        reaction_pb2.ProductCompound.Texture.AMORPHOUS_SOLID: 3,
+        reaction_pb2.ProductCompound.Texture.SEMI_SOLID: 3,
+        reaction_pb2.ProductCompound.Texture.SOLID: 3,
+    }
+    input_state_code = texture_type_to_state_of_matter[message.texture.type]
+    if input_state_code is not None:
+        components = message.components + message.crude_components
+        component_state_codes = [texture_type_to_state_of_matter[c.texture.type] for c in components]
+        if (
+            component_state_codes
+            and None not in component_state_codes
+            and max(component_state_codes) > input_state_code
```

Comment:
Can you document this a little more with in-line comments? My reading is that a warning will be issued if any component is solid but the input is liquid, which would be a common situation for stock solutions.

Reply:
my bad, it should have been `max(component_state_codes) < input_state_code` so warnings are issued for G->L (gas component, liquid reaction_input), L->S, GL->S, LL->S. I will add comments on this.